src/core/global.c

```diff
@@ -309,25 +308,33 @@ void nn_term (void)
     int i;
 
     nn_glock_lock ();
-
-    /*  Switch the global state into the zombie state. */
-    self.flags |= NN_CTX_FLAG_ZOMBIE;
-
-    /*  Mark all open sockets as terminating. */
-    if (self.nsocks != 0) {
-        for (i = 0; i != NN_MAX_SOCKETS; ++i)
-            if (self.socks [i] != NULL) {
-                nn_sock_zombify (self.socks [i]);
-            }
-    }
-
+    self.flags |= NN_CTX_FLAG_TERMING;
     nn_glock_unlock ();
 
     /* Make sure we really close resources, this will cause global
        resources to be freed too when the last socket is closed. */
     for (i = 0; i < NN_MAX_SOCKETS; i++) {
-        (void) nn_close_impl (i, 1);
+        (void) nn_close (i);
+    }
+
+    nn_glock_lock ();
+    self.flags |= NN_CTX_FLAG_TERMED;
+    self.flags &= ~NN_CTX_FLAG_TERMING;
+    /*  TODO: Add a signal to wake nn_init ()  */
+    nn_glock_unlock ();
+}
+
+void nn_init (void)
+{
+    nn_glock_lock ();
+    /*  Make sure any nn_term in progress is done. */
+    while (self.flags & NN_CTX_FLAG_TERMING) {
+        nn_glock_unlock ();
+        nn_sleep (100);
+        nn_glock_lock ();
     }
+    self.flags &= ~NN_CTX_FLAG_TERMED;
+    nn_glock_unlock ();
 }
```

Comment:
Is `nn_init` actually called anywhere right now? Or is this the TODO from line 323?